libs/libc/signal/sig_killpg.c

```diff
@@ -0,0 +1,69 @@
+/****************************************************************************
+ * libs/libc/signal/sig_killpg.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <errno.h>
+#include <signal.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: killpg
+ *
+ * Description:
+ *   The killpg() system call can be used to send any signal to process
+ *   group. See kill() for further information as this is just a simple
+ *   wrapper around the kill() function.
+ *
+ * Input Parameters:
+ *   pgrp  - The id of the process group to receive the signal.
+ *   signo - The signal number to send.  If 'signo' is zero, no signal is
+ *           sent, but all error checking is performed.
+ *
+ * Returned Value:
+ *    On success the signal was send and zero is returned. On error -1 is
+ *    returned, and errno is set one of the following error numbers:
+ *
+ *    EINVAL An invalid signal was specified.
+ *    EPERM  The thread does not have permission to send the
+ *           signal to the target thread.
+ *    ESRCH  No thread could be found corresponding to that
+ *           specified by the given thread ID
+ *    ENOSYS Do not support sending signals to process groups.
+ *
+ ****************************************************************************/
+
+int killpg(pid_t pgrp, int signo)
+{
+  if (pgrp <= 1)
+    {
+      set_errno(EINVAL);
```

Comment:
This case is not mentioned in the function description. Also I have a question about `1` since we mapped group pid to pid and `1` can be a valid pid number.

Reply:
according to the posix standard:https://pubs.opengroup.org/onlinepubs/9699919799/functions/killpg.html ,
"If pgrp is less than or equal to 1, the behavior of killpg() is undefined."
so we add judgement at here.

Reply:
Normally in Unix systems PID 1 is reserved to "init" process, it explains why killpg could be used against it.

Reply:
> Normally in Unix systems PID 1 is reserved to "init" process, it explains why killpg could be used against it.

The bring-up IDLE thread is PID 0.  The first task started will be PID 1.  In a single CPU build this will be the startup application (e.g., NSH).  It would be legitimate, I think to kill NSH.

However, we need a clear understanding of what is the init process: Is it the IDLE thread or NSH?  When a task daemonizes by killing its parent, the parent of the daemonized task should be the init process.  I think that has to be the IDLE thread PID 0.  I don't think that the IDLE thread has enough capability to do all of the things that a session group leader would need to do.

Linux also has a start up thread like NuttX, but it is not the init process.  It is invisible.

Things are more complicated in multi-CPU, SMP modes.  If, for example, there are 8 CPUs, then PIDs 0-7 are IDLE threads for each CPU.  None of these may be killed.

Reply:
> Is if "if (pgrp < NCPUs)" also valid for new MCUs that are using the concept of CPUs thread (I think it is like hyper-thread on PC) ? See: https://www.silabs.com/wireless/wi-fi/siwx917-wireless-socs

I have advocated for a long time that kernel threads are really very different from tasks and are (obviously super different from processes).  The really should not have most of the features of processes like file descriptors, streams (#1108) but they also should not have process IDs (thread IDs possibly, but not process IDs).

The fact that they do have process IDs results in illogical things like this.

NuttX has never been ported to a CPU with multiple cores each with multiple hardware threads, but I think that there would still be only a single startup IDLE thread for each CPU, rather than one per hardware thread.  If so, then the number of IDLE threads would still by NCPUs.